### DISCUSSION TITLE
How can I patch emscripten?

### DISCUSSION THREAD

**Discussion by km19809** [pyodide/pyodide#2978] [Q&A] [answered]
I am working on linking the pre-released version of SDL2 to pyodide.
It was successful, however, I need to patch some files of emscripten.
I don't know anything about patching, so I am changing files by myself.

What I want is to make a patch file, like [this](https://github.com/pyodide/pyodide/blob/main/emsdk/patches/0001-Add-BigInt64Array-shim-for-Safari-14.patch).

FYI, the diff I made:
```diff
diff --git a/tools/ports/sdl2.py b/tools/ports/sdl2.py
--- tools/ports/sdl2.py
+++ tools/ports/sdl2.py
@@ -4,10 +4,10 @@
 # found in the LICENSE file.
 
 import os
 
-TAG = '0fcfaf9e9482953ee425cc15b91160b92de3df7f'
-HASH = '2891b65bbe34ada80de6c83751b01cd259f877123f0eeb31cbacf8f7bac06b2157e17a2183f58911a9849ab57ca8aba165d5b3058ecd2d7191bfb61c0595807e'
+TAG = 'cb46e1b3f06a08d57b4ccd83127d1ec3139e1c0f'
+HASH = 'b245b89b81b15233d9db1663ebdfd864ce5dc3481ca974417f5168cc47d5d97861e90f49cfcf7801fe25a5e2a28a45ccf25b79b953e5e556a642eee6bda979a4'
 SUBDIR = 'SDL-' + TAG
 
 
 def needed(settings):
@@ -27,9 +27,9 @@
     source_include_path = os.path.join(ports.get_dir(), 'sdl2', SUBDIR, 'include')
     ports.install_headers(source_include_path, target='SDL2')
 
     # build
-    srcs = '''SDL.c SDL_assert.c SDL_dataqueue.c SDL_error.c SDL_guid.c SDL_hints.c SDL_list.c SDL_log.c
+    srcs = '''SDL.c SDL_assert.c SDL_dataqueue.c SDL_error.c SDL_guid.c SDL_hints.c SDL_list.c SDL_log.c SDL_utils.c 
     atomic/SDL_atomic.c atomic/SDL_spinlock.c audio/SDL_audio.c audio/SDL_audiocvt.c audio/SDL_audiodev.c
     audio/SDL_audiotypecvt.c audio/SDL_mixer.c audio/SDL_wave.c cpuinfo/SDL_cpuinfo.c
     dynapi/SDL_dynapi.c events/SDL_clipboardevents.c events/SDL_displayevents.c events/SDL_dropevents.c
     events/SDL_events.c events/SDL_gesture.c events/SDL_keyboard.c events/SDL_mouse.c events/SDL_quit.c
```

**Comment by ryanking13** [comment, marked as answer]
You can copy your patch files into emsdk/patches directory, [it will be applied automatically](https://github.com/pyodide/pyodide/blob/ed8f2b204e6e49d840d0c439b75f4d9893229363/emsdk/Makefile#L10) when building emsdk. Note that you need to call `make clean-all` then rerun `make` in order to rebuild emsdk.

**Reply by hoodmane** [reply]
The way I like to make patches is:
1. check out the version of emscripten (say 3.1.14) you want to patch
2. make a new branch
3. make changes
4. commit changes
5. `git format-patch 3.1.14 -o path/to/pyodide/emsdk/patches`

**Reply by km19809** [reply]
Thank you! I patched it!
